### PR TITLE
Switch to ESM

### LINKED ISSUE #238: Convert to ESM only
On 2021-04-30, 25 days from now, NodeJS 10 will reach end of life. From that point all NodeJS environments will support ESM. Of course module bundlers already support this.

There are two older pull requests to achieve this. Both are trying to keep backwards compatibility

- #150
- #194

I think it’s time to drop support for alternative module systems and bower, meaning no new devdependencies are needed.

Bower users can depend on specific git tags. It the `master` branch should keep working for Bower users, perhaps a new `main` branch should be created for further development, so the `master` branch can stay untouched indefinitely.

I expect this to get some backlash at first from people who don’t like breaking changes, but this is what major releases are for in semver.

### PR DESCRIPTION
**Pull request by jonkoops**
Closes #238

### CODE REVIEW

**Review by remcohaszing — APPROVED** [review]

**Comment by jonkoops** [comment]
Before merging this in we should agree on a strategy in #346.

**Comment by dcousens** [comment]
@jonkoops if we're compiling this far, why not Typescript native?

**Comment by dcousens** [comment]
I made an example project in https://github.com/dcousens/monorepo-typescript recently which I'm often using as a reference for this kind of project lift

**Comment by jonkoops** [comment]
> if we're compiling anyway, why not TypeScript native? Seeing as we're only 1 file deep for each of the exports, you won't need a bundler either

You are referring to Rollup in the `benchmarks` directory? I am actually looking to eliminate that entirely after this PR. In theory, all we need is a static file server to be able to import ESM code, which hopefully also eliminates any strange compiler quirks/overhead in the benchmarks.

I think a bundler/compiler (e.g Rollup/TypeScript would make sense if we decide to also support CommonJS next to ESM, but I am not sure if we would want to do so.

Perhaps we merge this in now, and make this a separate point of discussion/log an isssue? WDYT?

**Comment by dcousens** [comment]
> I am not sure if we would want to do so.
We might see slower uptake of `3.0.0` if we opt for ESM only

File: bind.js [hunk 1]
```diff
@@ -1,77 +1,54 @@
-/*!
-	Copyright (c) 2018 Jed Watson.
-	Licensed under the MIT License (MIT), see
-	http://jedwatson.github.io/classnames
-*/
-/* global define */
+var hasOwn = {}.hasOwnProperty;
```
Comment (dcousens) [inline bind.js:1]:
```suggestion
const hasOwn = {}.hasOwnProperty;
```

File: bind.js [hunk 2]
```diff
@@ -1,77 +1,54 @@
...
 
-(function () {
-	'use strict';
+export default function classNames () {
+	var classes = '';
```
Comment (dcousens) [inline bind.js:4]:
```suggestion
	let classes = '';
```

File: bind.js [hunk 3]
```diff
@@ -1,77 +1,54 @@
...
 
-	var hasOwn = {}.hasOwnProperty;
+	for (var i = 0; i < arguments.length; i++) {
+		var arg = arguments[i];
```
Comment (dcousens) [inline bind.js:7]:
```suggestion
	for (let i = 0; i < arguments.length; i++) {
		let arg = arguments[i];
```

File: bind.js [hunk 4]
```diff
@@ -1,77 +1,54 @@
...
-			return '';
-		}
+	if (arg.toString !== Object.prototype.toString && !arg.toString.toString().includes('[native code]')) {
+		return arg.toString();
+	}
 
-		if (Array.isArray(arg)) {
-			return classNames.apply(this, arg);
-		}
+	var classes = '';
```
Comment (dcousens) [inline bind.js:33]:
```suggestion
	let classes = '';
```

File: bind.js [hunk 5]
```diff
@@ -1,77 +1,54 @@
...
 
-		if (arg.toString !== Object.prototype.toString && !arg.toString.toString().includes('[native code]')) {
-			return arg.toString();
+	for (var key in arg) {
```
Comment (dcousens) [inline bind.js:35]:
```suggestion
	for (const key in arg) {
```

File: dedupe.js [hunk 6]
```diff
@@ -1,115 +1,94 @@
...
-		}
+var classNames = (function () {
+	// don't inherit from Object so we can skip hasOwnProperty check later
+	// http://stackoverflow.com/questions/15518328/creating-js-object-with-object-createnull#answer-21079232
+	function StorageObject() {}
+	StorageObject.prototype = Object.create(null);
 
-		var hasOwn = {}.hasOwnProperty;
+	function _parseArray (resultSet, array) {
+		var length = array.length;
```
Comment (dcousens) [inline dedupe.js:8]:
```suggestion
		const length = array.length;
```

File: dedupe.js [hunk 7]
```diff
@@ -1,115 +1,94 @@
...
+		resultSet[num] = true;
+	}
+
+	function _parseObject (resultSet, object) {
+		if (object.toString !== Object.prototype.toString && !object.toString.toString().includes('[native code]')) {
+			resultSet[object.toString()] = true;
+			return;
+		}
+
+		for (var k in object) {
```
Comment (dcousens) [inline dedupe.js:27]:
```suggestion
		for (const k in object) {
```

File: dedupe.js [hunk 8]
```diff
@@ -1,115 +1,94 @@
...
+	}
 
-		var SPACE = /\s+/;
-		function _parseString (resultSet, str) {
-			var array = str.split(SPACE);
-			var length = array.length;
+	var SPACE = /\s+/;
+	function _parseString (resultSet, str) {
+		var array = str.split(SPACE);
+		var length = array.length;
```
Comment (dcousens) [inline dedupe.js:39]:
```suggestion
	const SPACE = /\s+/;
	function _parseString (resultSet, str) {
		const array = str.split(SPACE);
		const length = array.length;
```

File: dedupe.js [hunk 9]
```diff
@@ -1,115 +1,94 @@
...
 
-			for (var i = 0; i < length; ++i) {
-				resultSet[array[i]] = true;
-			}
+		for (var i = 0; i < length; ++i) {
```
Comment (dcousens) [inline dedupe.js:41]:
```suggestion
		for (let i = 0; i < length; ++i) {
```

File: dedupe.js [hunk 10]
```diff
@@ -1,115 +1,94 @@
...
+			resultSet[array[i]] = true;
 		}
+	}
 
-		function _parse (resultSet, arg) {
-			if (!arg) return;
-			var argType = typeof arg;
+	function _parse (resultSet, arg) {
+		if (!arg) return;
+		var argType = typeof arg;
```
Comment (dcousens) [inline dedupe.js:48]:
```suggestion
		const argType = typeof arg;
```

File: dedupe.js [hunk 11]
```diff
@@ -1,115 +1,94 @@
...
-			var args = Array(len);
-			for (var i = 0; i < len; i++) {
-				args[i] = arguments[i];
-			}
+	function _classNames () {
+		// don't leak arguments
+		// https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#32-leaking-arguments
+		var len = arguments.length;
+		var args = Array(len);
+		for (var i = 0; i < len; i++) {
```
Comment (dcousens) [inline dedupe.js:73]:
```suggestion
		const length = arguments.length;
		const args = Array(len);
		for (const i = 0; i < length; i++) {
```

File: dedupe.js [hunk 12]
```diff
@@ -1,115 +1,94 @@
...
+			args[i] = arguments[i];
+		}
 
-			var classSet = new StorageObject();
-			_parseArray(classSet, args);
+		var classSet = new StorageObject();
+		_parseArray(classSet, args);
 
-			var list = [];
+		var list = [];
```
Comment (dcousens) [inline dedupe.js:80]:
```suggestion
		const list = [];
```

File: dedupe.js [hunk 13]
```diff
@@ -1,115 +1,94 @@
...
+		// https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#32-leaking-arguments
+		var len = arguments.length;
+		var args = Array(len);
+		for (var i = 0; i < len; i++) {
+			args[i] = arguments[i];
+		}
 
-			var classSet = new StorageObject();
-			_parseArray(classSet, args);
+		var classSet = new StorageObject();
```
Comment (dcousens) [inline dedupe.js:77]:
```suggestion
		const classSet = new StorageObject();
```

File: dedupe.js [hunk 14]
```diff
@@ -1,115 +1,94 @@
...
+		_parseArray(classSet, args);
 
-			var list = [];
+		var list = [];
 
-			for (var k in classSet) {
-				if (classSet[k]) {
-					list.push(k)
-				}
+		for (var k in classSet) {
```
Comment (dcousens) [inline dedupe.js:82]:
```suggestion
		for (const k in classSet) {
```

File: index.js [hunk 15]
```diff
@@ -1,77 +1,54 @@
-/*!
-	Copyright (c) 2018 Jed Watson.
-	Licensed under the MIT License (MIT), see
-	http://jedwatson.github.io/classnames
-*/
-/* global define */
+var hasOwn = {}.hasOwnProperty;
```
Comment (dcousens) [inline index.js:1]:
```suggestion
const hasOwn = {}.hasOwnProperty;
```

File: index.js [hunk 16]
```diff
@@ -1,77 +1,54 @@
...
 
-(function () {
-	'use strict';
+export default function classNames () {
+	var classes = '';
```
Comment (dcousens) [inline index.js:4]:
```suggestion
	let classes = '';
```

File: index.js [hunk 17]
```diff
@@ -1,77 +1,54 @@
...
 
-	var hasOwn = {}.hasOwnProperty;
+	for (var i = 0; i < arguments.length; i++) {
+		var arg = arguments[i];
```
Comment (dcousens) [inline index.js:7]:
```suggestion
	for (let i = 0; i < arguments.length; i++) {
		const arg = arguments[i];
```

File: index.js [hunk 18]
```diff
@@ -1,77 +1,54 @@
...
-			return '';
-		}
+	if (arg.toString !== Object.prototype.toString && !arg.toString.toString().includes('[native code]')) {
+		return arg.toString();
+	}
 
-		if (Array.isArray(arg)) {
-			return classNames.apply(null, arg);
-		}
+	var classes = '';
```
Comment (dcousens) [inline index.js:33]:
```suggestion
	let classes = '';
```

File: index.js [hunk 19]
```diff
@@ -1,77 +1,54 @@
...
 
-		if (arg.toString !== Object.prototype.toString && !arg.toString.toString().includes('[native code]')) {
-			return arg.toString();
+	for (var key in arg) {
```
Comment (dcousens) [inline index.js:35]:
```suggestion
	for (const key in arg) {
```

File: dedupe.js [hunk 20]
```diff
@@ -1,115 +1,94 @@
...
+			_parse(resultSet, array[i]);
 		}
+	}
 
-		function _parseObject (resultSet, object) {
-			if (object.toString !== Object.prototype.toString && !object.toString.toString().includes('[native code]')) {
-				resultSet[object.toString()] = true;
-				return;
-			}
+	var hasOwn = {}.hasOwnProperty;
```
Comment (dcousens) [inline dedupe.js:15]:
```suggestion
	const hasOwn = {}.hasOwnProperty;
```

File: dedupe.js [hunk 21]
```diff
@@ -1,115 +1,94 @@
...
+	function StorageObject() {}
+	StorageObject.prototype = Object.create(null);
 
-		var hasOwn = {}.hasOwnProperty;
+	function _parseArray (resultSet, array) {
+		const length = array.length;
 
-		function _parseNumber (resultSet, num) {
-			resultSet[num] = true;
+		for (var i = 0; i < length; ++i) {
```
Comment (dcousens) [inline dedupe.js:10]:
```suggestion
		for (let i = 0; i < length; ++i) {
```

File: dedupe.js [hunk 22]
```diff
@@ -1,115 +1,94 @@
...
-		StorageObject.prototype = Object.create(null);
-
-		function _parseArray (resultSet, array) {
-			var length = array.length;
-
-			for (var i = 0; i < length; ++i) {
-				_parse(resultSet, array[i]);
-			}
-		}
+var classNames = (function () {
```
Comment (dcousens) [inline dedupe.js:1]:
```suggestion
const classNames = (function () {
```

File: dedupe.js [hunk 23]
```diff
@@ -1,115 +1,94 @@
...
-			var args = Array(len);
-			for (var i = 0; i < len; i++) {
-				args[i] = arguments[i];
-			}
+	function _classNames () {
+		// don't leak arguments
+		// https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#32-leaking-arguments
+		const length = arguments.length;
+		const args = Array(len);
+		for (const i = 0; i < length; i++) {
```
Comment (dcousens) [inline dedupe.js:73]:
```suggestion
		for (let i = 0; i < length; i++) {
```

**Comment by remcohaszing** [comment]
> I think a bundler/compiler (e.g Rollup/TypeScript would make sense if we decide to also support CommonJS next to ESM, but I am not sure if we would want to do so.

Please don't do this. Although many people do so, TypeScript isn't intended to be used for dual publishing. Especially the combination of TypeScript, default exports, and Rollup is a well-known recipe for generating invalid type definitions.

**Comment by jonkoops** [comment]
Let's continue the CommonJS/TypeScript discussion under #349, and focus on landing this work here. I have some other things in the pipeline that depend on it, such as removing the need for a bundler from the benchmarks.

**Comment by jonkoops** [comment]
@dcousens WDYT? Can we merge this in?

**Pull request merged**